### PR TITLE
Add support for IceCap

### PR DESCRIPTION
**Pull request by nspin**
This PR adds support for [IceCap](https://gitlab.com/arm-research/security/icecap/icecap) as an isolation platform for Veracruz.

This PR accompanies https://github.com/veracruz-project/veracruz/pull/149.

At the time of writing, this is a draft, submitted for the purpose of starting a discussion and gathering feedback.

### CODE REVIEW

**Comment by ericvh** [comment]
> In the case where `/nix` is mounted as a volume, `/nix` is not available at image build time.

Yeah, I get that -- the solution I was thinking of to this for MathiasBr's cache (which has the same problem is an additional Makefile rule which populates the local cache from the Docker (if you are using one and its not already there).  To be fair I don't have any analysis of why this would be any better/worse -- I guess it depends on how much "building" is being done we might be able to cache in the docker image versus have to wait for on the host.  I have built your Docker but I haven't tried building the contents yet, so maybe I'll have better insight then -- but when looking at incorporating MathiasBr's caching I was looking for the right middle ground that would keep CI runs tractable on build time so trying to balance pre-build of pre-requisits versus docker image size.

**Comment by ericvh** [comment]
> I agree. I think we should.
> 
> I think that it would be good to do this in a way where users who don't want to learn anything about Nix don't have to make a decision. That is, I think this should be an "advanced" configuration option.

In many ways I think the local cache best benefits folks who are doing active development on the thing being cached and/or might be using that same environment for something besides veracruz.  So, for example, cargo/rust might be more widely sharable between Dom's local linux veracruz test platform and the target platform.  For Nix the same thing may be true for anyone that uses Nix in the rest of their environment.  Otherwise I think I just want to pull the image -- and even if I want to have a local cache it may still be faster to pull the image and copy to the local cache.  But I agree that the latter is the "advanced" option.  So, summary - I "think" docker-only pre-reqs is probably the easy option and we configure cache with environment variables and/or slightly different make arguments.

> 
> What do you think the default should be? I see pros of both.

See my previous comment, Iet me do a run-through of actually building icecap in this way and then when I look at image size versus compilation time I'll have a better informed opinion.

My starting position is that anything that isn't veracruz itself can (and probably should) be pre-built -- but depending on resulting image sizes, etc. I may bow to a more practical/pragmatic solution.  The big thing for me is, (assumption: if we are using generic github actions runners) CI is going to be rebuilding every time from scratch and so I want as little rebuild time as possible.

**Comment by nspin** [comment]
> > In the case where `/nix` is mounted as a volume, `/nix` is not available at image build time.
> 
> Yeah, I get that -- the solution I was thinking of to this for MathiasBr's cache (which has the same problem is an additional Makefile rule which populates the local cache from the Docker (if you are using one and its not already there). To be fair I don't have any analysis of why this would be any better/worse -- I guess it depends on how much "building" is being done we might be able to cache in the docker image versus have to wait for on the host. I have built your Docker but I haven't tried building the contents yet, so maybe I'll have better insight then -- but when looking at incorporating MathiasBr's caching I was looking for the right middle ground that would keep CI runs tractable on build time so trying to balance pre-build of pre-requisits versus docker image size.

I'll address the wider issue of to bind mount or not to bind mound in response to your other comment. As for this entry-point script, Nix installation is just downloading, unpacking, and copying 43M. We could do the downloading at image build time.

**Comment by nspin** [comment]
> My starting position is that anything that isn't veracruz itself can (and probably should) be pre-built -- but depending on resulting image sizes, etc. I may bow to a more practical/pragmatic solution. The big thing for me is, (assumption: if we are using generic github actions runners) CI is going to be rebuilding every time from scratch and so I want as little rebuild time as possible.

That sounds like a good idea to me.

In doing so, we run the risk of ending up in a situation where trivial changes cause very long image rebuild times. One way to avoid this would be a remote Nix binary cache (which is already incorporated into https://github.com/veracruz-project/veracruz/pull/149), if we deemed maintaining one worth the effort.

File: Dockerfile [hunk 1]
```diff
@@ -310,10 +310,29 @@ RUN chown -R $USER /work/rust-optee-trustzone-sdk/optee-qemuv8-3.7.0/build/build
 WORKDIR /
 RUN ./cleanup.sh
 
+FROM build_sgx as build_icecap
```
Comment (dreemkiller) [inline Dockerfile:313]: Do we want to use the sgx build as the base for icecap? Doesn't it have things we don't need for icecap?

File: Dockerfile [hunk 2]
```diff
@@ -310,10 +310,29 @@ RUN chown -R $USER /work/rust-optee-trustzone-sdk/optee-qemuv8-3.7.0/build/build
...
+USER $USER
+
+ENV PATH="/nix/env/bin:${PATH}"
+ENV MANPATH="/nix/env/share/man:${MANPATH}"
+ENV NIX_SSL_CERT_FILE=/nix/env/etc/ssl/certs/ca-bundle.crt
+
+# HACK for persistent git cache. $XDG_CACHE_HOME doesn't appear to be in use elsewhere.
+# TODO patch nix to use $NIX_CACHE_HACK instead
+ENV XDG_CACHE_HOME=/nix/cache
+
```
Comment (dreemkiller) [inline Dockerfile:331]: Wouldn't it be better to have these in an Icecap-specific Dockerfile, similar to the rest of the supported platforms?

Comment (nspin) [reply Dockerfile:331]: I still haven't rebased this on top of Eric's splitting. Doing so will address this comment and you other comment about SGX as a base.

**Comment by dominic-mulligan-arm** [comment]
What is the status of this PR, @nspin?  Can it be closed?

**Comment by nspin** [comment]
Yes. Someday we may move the IceCap Docker setup into this repository, but I think we shouldn't worry about that until after our upcoming deadlines pass.

**Comment by dominic-mulligan-arm** [comment]
Closed after confirmation from @nspin.

**Pull request closed without merging**